File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        q = deque()
        q.append((root,0))
        maxlen = 0
        while q:
            size = len(q)
            first = q[0][1]
            last = q[-1][1]
            maxlen = max(maxlen, last - first  + 1) 
            for _ in range(size):
                node, index = q.popleft()
                if node.left:
                    q.append((node.left, index*2))
                if node.right:
                    q.append((node.right, index*2 + 1))
        
        return maxlen

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if(root==NULL){
            return 0;
        }

        queue<pair<TreeNode*, long long>> q;
        long long maxwidth = 0;
        q.push({root, 1});
        while(!q.empty()){
            int size = q.size();
            long long first = q.front().second;
            long long last = q.back().second;
            maxwidth = max(maxwidth, abs(last - first + 1));
            while(size--){
                auto node= q.front().first;
                long long index = q.front().second - first;
                q.pop();
                
                if(node->left) q.push({node->left, 2*index});
                if(node->right) q.push({node->right, 2*index + 1});
            }
        }

        

        return maxwidth;

    }
};